README.md
[![](https://img.shields.io/pypi/v/cir-model)](https://pypi.org/project/cir-model/) [![License](https://img.shields.io/badge/License-BSD_3--Clause-blue.svg)](https://github.com/mathijs02/cir-model/blob/main/LICENSE.txt)

# Introduction
The `cir-model` library implements the *Centered Isotonic Regression* (CIR) model.[1] CIR is a variant of [Isotonic Regression](https://en.wikipedia.org/wiki/Isotonic_regression) (IR), which is a nonparametric regression model which only assumes that the data generating function is monotonically increasing or decreasing. The function can take any monotonic form that minimizes the sum of squared errors. [`scikit-learn`](https://scikit-learn.org/stable/) has implemented Isotonic Regression as [`IsotonicRegression`](https://scikit-learn.org/stable/modules/generated/sklearn.isotonic.IsotonicRegression.html).[2]

In some scenarios, the assumption of monotonicity is too weak. IR can result in a function that has constant, non-decreasing intervals. In some applications, like dose-response curve fitting or in cumululative distribution function estimation, it can be safe to make the additional assumption of *strict monotonicity*: at any point the function should be increasing, not just non-decreasing (or vice versa: decreasing, not just non-increasing). This additional requirement results in a function with the desirable property that the inverse function is a unique mapping, because there are no longer ranges of constant intervals.

*Centered Isotonic Regression* (CIR) is similar to IR, but with the additional assumption of strict monotonicity. The algorithm is described in detail in [1]. The plot below shows a comparison of IR (in blue) and CIR (in red) fitted to the same datapoints (in green).

![](https://github.com/mathijs02/cir-model/raw/main/examples/ir_cir_comparison.png)

This Python library, `cir-model`, implements `CenteredIsotonicRegression` in line with the algorithm in [1]. `cir-model` uses the `IsotonicRegression` implementation of `scikit-learn`. Therefore, `CenteredIsotonicRegression` takes the same parameters as `IsotonicRegression` and is completely compatible with `scikit-learn`. You can use it for example in `scikit-learn` pipelines.

You can install `cir-model` using `pip`:
```
pip install -U cir-model
```

# Examples
To fit a CIR model:
```python
>>> from cir_model import CenteredIsotonicRegression
>>> x = [1, 2, 3, 4]
>>> y = [1, 37, 42, 5]
>>> model = CenteredIsotonicRegression().fit(x, y)
>>> model.transform(x)
array([ 1. , 14.5, 28. , 28. ])
```

Finding the inverse of the CIR model above, for example for the value `x` for which `y=25`:
```python
>>> from scipy import optimize
>>> optimize.newton(lambda x: model.transform([x]) - 25, 2)
array([2.77777778])
```

# References
1. [Centered Isotonic Regression: Point and Interval Estimation for Dose-Response Studies](https://arxiv.org/abs/1701.05964), Assaf P. Oron & Nancy Flournoy, Statistics in Biopharmaceutical Research, Volume 9, Issue 3, 258-267, 2017
2. [Scikit-learn: Machine Learning in Python](https://jmlr.csail.mit.edu/papers/v12/pedregosa11a.html), Fabian Pedregosa et al., JMLR 12, 2825-2830, 2011

CHANGELOG.md
# Changelog

## v0.2.0
* Added the `non_centered_points` parameter.

## v0.1.0
* First release, implementing Centered Isotonic Regression (CIR) in line with algorithm in original CIR paper.

tests/test_cir_model.py
"""
This module tests the CenteredIsotonicRegression class.
"""

import itertools
import pickle
from typing import List, Union

import numpy as np
import pytest
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler

from cir_model import CenteredIsotonicRegression

EXAMPLE_X = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
EXAMPLE_Y = [1, 37, 42, 5, 7, 14, 27, 32, 56, 43]
EXAMPLE_WT = [1, 1, 2, 2, 3, 3, 1, 1, 4, 4]
EXAMPLE_OUTPUT_WT = [
    1.0,
    5.81578947,
    10.63157895,
    15.44736842,
    19.64285714,
    23.32142857,
    27.0,
    32.0,
    43.66666667,
    49.5,
]
EXAMPLE_OUTPUT_NO_WT = [
    1.0,
    7.66666667,
    14.33333333,
    21.0,
    23.0,
    25.0,
    27.0,
    32.0,
    43.66666667,
    49.5,
]


@pytest.mark.parametrize(
    "increasing, expected",
    [
        ("auto", EXAMPLE_OUTPUT_WT),
        (True, EXAMPLE_OUTPUT_WT),
        (
            False,
            [
                30.5,
                30.5,
                30.32786885,
                30.09836066,
                29.86885246,
                29.63934426,
                29.40983607,
                29.33333333,
                29.33333333,
                29.33333333,
            ],
        ),
    ],
)
def test_cir_increasing(
    increasing: Union[str, bool],
    expected: List[float],
) -> None:
    """
    Test whether the expected prediction values are obtained for various values
    of `increasing`.
    """
    cir = CenteredIsotonicRegression(increasing=increasing)
    output = cir.fit_transform(EXAMPLE_X, EXAMPLE_Y, sample_weight=EXAMPLE_WT)
    np.testing.assert_array_almost_equal(expected, output)


def test_cir_fit_transform() -> None:
    """
    Test whether the same prediction values are obtained when using `fit` and
    `transform` compared to `fit_transform`.
    """
    output_f = (
        CenteredIsotonicRegression()
        .fit(EXAMPLE_X, EXAMPLE_Y, sample_weight=EXAMPLE_WT)
        .transform(EXAMPLE_X)
    )
    output_ft = CenteredIsotonicRegression().fit_transform(
        EXAMPLE_X, EXAMPLE_Y, sample_weight=EXAMPLE_WT
    )
    np.testing.assert_array_almost_equal(output_f, output_ft)


def test_cir_one_value() -> None:
    """
    Test the fitting when the training dataset contains only one datapoint.
    """
    output = CenteredIsotonicRegression(increasing="auto").fit_transform(
        [1], [5], sample_weight=[2]
    )
    np.testing.assert_array_almost_equal(output, [5])


def test_cir_no_weights() -> None:
    """
    Test the fitting when no weights are given.
    """
    output = (
        CenteredIsotonicRegression()
        .fit(EXAMPLE_X, EXAMPLE_Y)
        .transform(EXAMPLE_X)
    )
    np.testing.assert_array_almost_equal(output, EXAMPLE_OUTPUT_NO_WT)


def test_cir_no_weights_equal_ones() -> None:
    """
    Test that passing no weights gives the same output as passing ones as
    weights.
    """
    no_wt = CenteredIsotonicRegression().fit_transform(EXAMPLE_X, EXAMPLE_Y)
    wt_ones = CenteredIsotonicRegression().fit_transform(
        EXAMPLE_X, EXAMPLE_Y, sample_weight=np.ones(len(EXAMPLE_X))
    )
    np.testing.assert_array_almost_equal(no_wt, wt_ones)


@pytest.mark.parametrize("order", map(list, itertools.permutations(range(4))))
def test_cir_all_permutations(order: List[int]) -> None:
    """
    Test whether the outcome is independent of the ordering in the training
    dataset by testing all possible permutations of a small dataset.
    """
    print(order)
    x = np.array([1, 2, 3, 4])[order]
    y = np.array([1, 37, 42, 5])[order]
    wt = np.array([1, 1, 2, 2])[order]

    cir = CenteredIsotonicRegression()
    cir.fit(x, y, wt)

    output = cir.transform([1, 2, 3, 4])
    np.testing.assert_array_almost_equal(
        [1.0, 12.45454545, 23.90909091, 26.2], output
    )


def test_cir_1d_array() -> None:
    """
    Test whether numpy arrays as inputs work as expected.
    """
    cir = CenteredIsotonicRegression()
    output = cir.fit_transform(
        np.array(EXAMPLE_X), np.array(EXAMPLE_Y), sample_weight=EXAMPLE_WT
    )
    np.testing.assert_array_almost_equal(EXAMPLE_OUTPUT_WT, output)


def test_cir_2d_array() -> None:
    """
    Test whether numpy arrays as input, with a 2D array for X, works as
    expected.
    """
    cir = CenteredIsotonicRegression()
    output = cir.fit_transform(
        np.array([EXAMPLE_X]).T, np.array(EXAMPLE_Y), sample_weight=EXAMPLE_WT
    )
    np.testing.assert_array_almost_equal(EXAMPLE_OUTPUT_WT, output)


def test_cir_incompatible_shapes() -> None:
    """
    Test whether an error is thrown when incompatible array sizes are passed.
    """
    cir = CenteredIsotonicRegression()
    with pytest.raises(ValueError):
        cir.fit(EXAMPLE_X, EXAMPLE_Y, EXAMPLE_WT[:-2])


def test_cir_pickle() -> None:
    """
    Test whether pickling and unpickling a model results in the same
    predictions.
    """
    cir = CenteredIsotonicRegression()
    cir.fit(EXAMPLE_X, EXAMPLE_Y, EXAMPLE_WT)
    original_output = cir.transform(EXAMPLE_X)

    cir_pickled = pickle.dumps(cir)
    cir_unpickled = pickle.loads(cir_pickled)
    pickled_output = cir_unpickled.transform(EXAMPLE_X)

    np.testing.assert_array_almost_equal(original_output, pickled_output)


def test_cir_oob_nan() -> None:
    """
    Test the intended behaviour of `out_of_bounds="nan"`.
    """
    cir = CenteredIsotonicRegression(out_of_bounds="nan")
    cir.fit(EXAMPLE_X, EXAMPLE_Y, EXAMPLE_WT)
    output = cir.transform(np.array([-1, 0, 1, 2]))
    expected = np.array([np.nan, np.nan, 1.0, 5.81578947])
    np.testing.assert_array_almost_equal(expected, output)


def test_cir_oob_clip() -> None:
    """
    Test the intended behaviour of `out_of_bounds="clip"`.
    """
    cir = CenteredIsotonicRegression(out_of_bounds="clip")
    cir.fit(EXAMPLE_X, EXAMPLE_Y, EXAMPLE_WT)
    output = cir.transform(np.array([-1, 0, 1, 2]))
    expected = np.array([1.0, 1.0, 1.0, 5.81578947])
    np.testing.assert_array_almost_equal(expected, output)


def test_cir_oob_raise() -> None:
    """
    Test the intended behaviour of `out_of_bounds="raise"`.
    """
    cir = CenteredIsotonicRegression(out_of_bounds="raise")
    cir.fit(EXAMPLE_X, EXAMPLE_Y, EXAMPLE_WT)
    with pytest.raises(ValueError):
        cir.transform(np.array([-1, 0, 1, 2]))


def test_cir_pipeline() -> None:
    """
    Test whether CenteredIsotonicRegression is compatible with scikit-learn
    pipelines.
    """
    x = np.array([EXAMPLE_X]).T
    y = np.array(EXAMPLE_Y)
    wt = np.array(EXAMPLE_WT)

    pipe = Pipeline(
        [("scaler", StandardScaler()), ("cir", CenteredIsotonicRegression())]
    )
    pipe.fit(x, y, cir__sample_weight=wt)
    output = pipe.transform(x)
    np.testing.assert_array_almost_equal(EXAMPLE_OUTPUT_WT, output)


@pytest.mark.parametrize(
    "y_min, y_max, expected",
    [
        (None, None, EXAMPLE_OUTPUT_WT),
        (0.5, 50, EXAMPLE_OUTPUT_WT),
        (
            3,
            50,
            [
                3.0,
                7.23684211,
                11.47368421,
                15.71052632,
                19.64285714,
                23.32142857,
                27.0,
                32.0,
                43.66666667,
                49.5,
            ],
        ),
        (
            25,
            50,
            [
                25.0,
                25.0,
                25.0,
                25.0,
                25.58823529,
                26.29411765,
                27.0,
                32.0,
                43.66666667,
                49.5,
            ],
        ),
        (
            3,
            None,
            [
                3.0,
                7.23684211,
                11.47368421,
                15.71052632,
                19.64285714,
                23.32142857,
                27.0,
                32.0,
                43.66666667,
                49.5,
            ],
        ),
        (
            25,
            None,
            [
                25.0,
                25.0,
                25.0,
                25.0,
                25.58823529,
                26.29411765,
                27.0,
                32.0,
                43.66666667,
                49.5,
            ],
        ),
        (
            0.5,
            45,
            [
                1.0,
                5.81578947,
                10.63157895,
                15.44736842,
                19.64285714,
                23.32142857,
                27.0,
                32.0,
                40.66666667,
                45.0,
            ],
        ),
        (
            0.5,
            25,
            [
                1.0,
                5.81578947,
                10.63157895,
                15.44736842,
                18.50097847,
                20.08610568,
                21.67123288,
                23.25636008,
                24.84148728,
                25.0,
            ],
        ),
        (
            None,
            45,
            [
                1.0,
                5.81578947,
                10.63157895,
                15.44736842,
                19.64285714,
                23.32142857,
                27.0,
                32.0,
                40.66666667,
                45.0,
            ],
        ),
        (
            None,
            25,
            [
                1.0,
                5.81578947,
                10.63157895,
                15.44736842,
                18.50097847,
                20.08610568,
                21.67123288,
                23.25636008,
                24.84148728,
                25.0,
            ],
        ),
        (
            3,
            45,
            [
                3.0,
                7.23684211,
                11.47368421,
                15.71052632,
                19.64285714,
                23.32142857,
                27.0,
                32.0,
                40.66666667,
                45.0,
            ],
        ),
        (
            22,
            35,
            [
                22.0,
                22.0,
                22.0,
                22.0,
                23.47058824,
                25.23529412,
                27.0,
                32.0,
                34.0,
                35.0,
            ],
        ),
    ],
)
def test_cir_ymin_ymax(
    y_min: float,
    y_max: float,
    expected: List[float],
) -> None:
    """
    Test whether the `y_min` and `y_max` functionality from
    `IsotonicRegression` works as expected with `CenteredIsotonicRegression`.
    """
    cir = CenteredIsotonicRegression(y_min=y_min, y_max=y_max)
    output = cir.fit_transform(EXAMPLE_X, EXAMPLE_Y, sample_weight=EXAMPLE_WT)
    np.testing.assert_array_almost_equal(expected, output)


@pytest.mark.parametrize(
    "y, non_centered_points, expected",
    [
        ([-1, -0.5, 0, 0, 0.5, 1], [0, 1], [-1.0, -0.5, 0.0, 0.0, 0.5, 1.0]),
        (
            [-1, -0.5, 0, 0, 0.5, 1],
            [1],
            [-1.0, -0.5, -0.166667, 0.166667, 0.5, 1.0],
        ),
        (
            [-1, -0.5, 0, 0, 0.5, 1],
            [],
            [-1.0, -0.5, -0.166667, 0.166667, 0.5, 1.0],
        ),
        (
            [-1, -0.5, 0.2, 0.2, 0.5, 1],
            [0.2],
            [-1.0, -0.5, 0.2, 0.2, 0.5, 1.0],
        ),
        (
            [-1, -0.5, 0.2, 0.2, 0.5, 1],
            [],
            [-1.0, -0.5, -0.03333333, 0.30000000, 0.5, 1.0],
        ),
        (
            [0, 0, 0, 2, 1.5, 2.7],
            [0],
            [0.0, 0.0, 0.0, 1.16666667, 2.06666667, 2.7],
        ),
    ],
)
def test_cir_non_centered_points(
    y: List[float],
    non_centered_points: List[float],
    expected: List[float],
) -> None:
    """
    Test whether the `non_centered_points` parameter works as expected.
    """
    x = [1, 2, 3, 4, 5, 6]
    cir = CenteredIsotonicRegression(non_centered_points=non_centered_points)
    output = cir.fit_transform(x, y)
    np.testing.assert_array_almost_equal(expected, output)


def test_cir_non_centered_points_missing() -> None:
    """
    Test whether the default value of the `non_centered_points` parameter is
    correct.
    """
    x = [1, 2, 3, 4, 5, 6]
    y = [-1, -0.5, 0, 0, 0.5, 1]
    cir_missing = CenteredIsotonicRegression().fit_transform(x, y)
    cir_default = CenteredIsotonicRegression(
        non_centered_points=[0, 1]
    ).fit_transform(x, y)
    np.testing.assert_array_almost_equal(cir_missing, cir_default)


examples/benchmark.py
"""
This module runs a benchmark of IsotonicRegression and
CenteredIsotonicRegression, to understand how they are different in regards to
computational resources.
"""

import logging
import time
from datetime import timedelta
from typing import Tuple

import numpy as np
from scipy.stats import logistic
from sklearn.isotonic import IsotonicRegression

from cir_model import CenteredIsotonicRegression

logging.basicConfig(level=logging.INFO)


def generate_data(n: int) -> Tuple[np.ndarray, np.ndarray]:
    """
    Generate an artifical dataset with a noisy sigmoid data generating
    function.
    """
    np.random.seed(10)
    x = np.sort(np.random.uniform(0, 35, n))
    noise = np.random.normal(0, 2, n)
    y = logistic.cdf(-x / 3 + 5 + noise)
    return x, y


def benchmark_ir_cir(n_datapoints: int, n_repeats: int, n_cycles: int) -> None:
    """
    Run a benchmark of CenteredIsotonicRegression against IsotonicRegression.
    The model fitting is repeated, to obtain less noisy estimates, and it
    cycles multiple times through the models to obtain a fair comparison: the
    first run is typically slower.
    """
    x, y = generate_data(n_datapoints)

    for model in n_cycles * [IsotonicRegression, CenteredIsotonicRegression]:
        start_time = time.monotonic()
        for _ in range(n_repeats):
            model().fit_transform(x, y)

        end_time = time.monotonic()
        time_diff = timedelta(seconds=end_time - start_time)
        logging.info(f"Time taken by {model}: {time_diff}")


if __name__ == "__main__":
    benchmark_ir_cir(100000, 500, 3)


examples/ir_cir_comparison.py
"""
This module shows an example usage of `CenteredIsotonicRegression` and saves a
plot comparing IR and CIR models fitted on the same data.
"""

import matplotlib.pyplot as plt
from benchmark import generate_data
from sklearn.isotonic import IsotonicRegression

from cir_model import CenteredIsotonicRegression

x, y = generate_data(100)
ir = IsotonicRegression(increasing=False).fit(x, y)
cir = CenteredIsotonicRegression(increasing=False).fit(x, y)

fig, ax = plt.subplots(figsize=(4.5, 3))
ax.scatter(x, y, color="C2", alpha=0.2, label="data")
ax.plot(
    ir.f_.x,
    ir.f_.y,
    color="C0",
    marker="o",
    markersize=5,
    label="IR",
    alpha=0.6,
)
ax.plot(
    cir.f_.x,
    cir.f_.y,
    color="C3",
    marker="o",
    markersize=5,
    label="CIR",
    alpha=0.6,
)
ax.legend()

plt.savefig("ir_cir_comparison.png")


src/cir_model/cir_model.py
"""
This module contains the CenteredIsotonicRegression class.
"""

from typing import Any, List, Optional, Tuple, Union

import numpy as np
from sklearn.isotonic import IsotonicRegression


class CenteredIsotonicRegression(IsotonicRegression):
    """
    Centered Isotonic Regression (CIR) model. CIR is described in [1]_ and is
    similar to Isotonic Regression (IR). CIR takes as an additional constraint,
    compared to IR, that the resulting function needs to be strictly monotonic:
    ranges of constant function values are prevented as much as possible.
    The `CenteredIsotonicRegression` class inherits all methods and attributes
    from the `scikit-learn` implementation `IsotonicRegression` and it is
    therefore compatible with the other components of the `scikit-learn`
    library, like for example pipelines.

    Parameters
    ----------
    This class takes the same parameters and has the same attributes as
    `IsotonicRegression` from `scikit-learn`.[2]_ For full documentation of
    `IsotonicRegression`, see:
    https://scikit-learn.org/stable/modules/generated/sklearn.isotonic.IsotonicRegression.html

    CenteredIsotonicRegression takes one additional parameter:

    non_centered_points : list, default: [0, 1]
        A list of y values that should not be collapsed in the CIR algorithm.
        In the original CIR algorithm, y values of 0 and 1 are treated
        differently by not collapsing them. This is because CIR is typically
        used for a binary target variable. The default behaviour can be
        overruled by passing a list of values for `non_centered_points`. An
        empty list means that no points are treated differently.

    References
    ----------
    .. [1] Centered Isotonic Regression: Point and Interval Estimation for
           Dose-Response Studies, Assaf P. Oron & Nancy Flournoy, Statistics in
           Biopharmaceutical Research, Volume 9, Issue 3, 258-267, 2017

    .. [2] Scikit-learn: Machine Learning in Python, Fabian Pedregosa et al.,
           JMLR 12, 2825-2830, 2011

    Examples
    --------
    >>> from cir_model import CenteredIsotonicRegression
    >>> x = [1, 2, 3, 4]
    >>> y = [1, 21, 41, 34]
    >>> model = CenteredIsotonicRegression().fit(x, y)
    >>> model.transform(x)
    array([ 1. , 21. , 32. , 37.5])
    """

    def __init__(
        self,
        non_centered_points: List[Union[float, int]] = [0, 1],
        **kwargs: Any,
    ) -> None:
        super().__init__(**kwargs)
        self.non_centered_points = non_centered_points

    def fit(
        self,
        X: Union[np.ndarray, List],
        y: Union[np.ndarray, List],
        sample_weight: Optional[Union[np.ndarray, List]] = None,
    ) -> "CenteredIsotonicRegression":
        """
        Fit the model using X, y and optionally sample_weight as training data.
        This method takes the same parameters and returns the same objects as
        `fit` from `IsotonicRegression`. For full documentation of
        `IsotonicRegression`, see:
        https://scikit-learn.org/stable/modules/generated/sklearn.isotonic.IsotonicRegression.html#sklearn.isotonic.IsotonicRegression.fit
        """

        super().fit(X, y, sample_weight)
        x_new, y_new = self._build_cir_points(X, sample_weight)

        # In order to create the final model, here is a second call to the IR
        # `fit` method. This might suggest that `CenteredIsotonicRegression` is
        # at least twice as slow as `IsotonicRegression``, but the second call
        # is typically very light because there are typically much fewer
        # datapoints. Also, the datapoints are already monotonic. A benchmark
        # on a realistic dataset shows that `CenteredIsotonicRegression` is
        # about 15% slower than `IsotonicRegression`.
        #
        # The advantage of calling `fit` a second time, instead of rebuilding
        # only the prediction function `self.f_`, is that in this way there are
        # no calls to non-public methods of `IsotonicRegression` and therefore
        # this maximizes compatibility with `scikit-learn`.
        super().fit(x_new, y_new, np.ones(x_new.shape))
        return self

    def _build_cir_points(
        self,
        X: Union[np.ndarray, List],
        sample_weight: Optional[Union[np.ndarray, List]],
    ) -> Tuple[np.ndarray, np.ndarray]:
        """
        Generate new x, y points for the interpolation function of Isotonic
        Regression, in line with the definition of the Centered Isotonic
        Regression model. We use here that the CIR points can be constructed
        using the outcome of IR and the training data.

        * To transform a trained IR model to a CIR model, every range of
          constant values in the interpolation function is collapsed into one
          point with as x-coordinate the weighted average of the training
          datapoints within the constant range.
        * In the CIR paper, ranges with constant values of 0 or 1 are not
          collapsed. This implementation takes the non-collapsible points as a
          parameter.
        * The original range of IR is kept in CIR. This means that ranges of
          constant values can appear at the edges of the function's domain.
        """
        points_new = []

        # Input parameters were already validated by calling `super().fit`
        X_arr = np.array(X).reshape(-1)

        order = np.argsort(X_arr)
        X_arr = X_arr[order]
        sample_weight_arr = (
            np.ones(X_arr.shape)
            if sample_weight is None
            else np.array(sample_weight)[order]
        )

        y = self.transform(X_arr)
        _, idx = np.unique(y, return_index=True)
        y_steps = y[np.sort(idx)]

        for n, y_step in enumerate(y_steps):
            idx = np.where(y == y_step)[0]
            x_step = X_arr[idx]
            x_mean = np.average(x_step, weights=sample_weight_arr[idx])

            # Points that should not be collapsed
            if y_step in self.non_centered_points:
                points_new.extend([(x_step[0], y_step), (x_step[-1], y_step)])
            # Ensure that the original range is maintained
            elif n == 0:
                points_new.extend([(x_step[0], y_step), (x_mean, y_step)])
            elif n == len(y_steps) - 1:
                points_new.extend([(x_mean, y_step), (x_step[-1], y_step)])
            else:
                points_new.append((x_mean, y_step))

        x_new, y_new = np.array(list(set(points_new))).T
        return x_new, y_new
